File: src/minigames/speed_builders/java/work/lclpnet/ap2/game/speed_builders/util/SbManager.java
package work.lclpnet.ap2.game.speed_builders.util;

import it.unimi.dsi.fastutil.objects.Object2LongMap;
import it.unimi.dsi.fastutil.objects.Object2LongOpenHashMap;
import lombok.Setter;
import net.minecraft.ChatFormatting;
import net.minecraft.core.BlockPos;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.server.level.ServerPlayer;
import net.minecraft.server.players.PlayerList;
import net.minecraft.sounds.SoundEvents;
import net.minecraft.sounds.SoundSource;
import net.minecraft.world.entity.Entity;
import net.minecraft.world.entity.player.Inventory;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.item.Items;
import net.minecraft.world.scores.PlayerTeam;
import org.jetbrains.annotations.NotNull;
import org.slf4j.Logger;
import work.lclpnet.ap2.api.base.Participants;
import work.lclpnet.ap2.api.game.MiniGameHandle;
import work.lclpnet.ap2.game.speed_builders.data.SbIsland;
import work.lclpnet.ap2.game.speed_builders.data.SbModule;
import work.lclpnet.ap2.impl.util.scoreboard.CustomScoreboardManager;
import work.lclpnet.kibu.access.entity.ServerPlayerAccess;

import java.util.*;
import java.util.function.BiConsumer;
import java.util.stream.Collectors;

public class SbManager {

    private static final int BASE_BUILD_DURATION_SECONDS = 25;
    private static final int MIN_BUILD_DURATION_SECONDS = 5;
    private static final int SUCCESSIVE_COMPLETION_REDUCTION_SECONDS = 7;

    private final Map<UUID, SbIsland> islands;
    private final List<SbModule> modules;
    private final MiniGameHandle gameHandle;
    private final ServerLevel world;
    private final Logger logger;
    private final Random random;
    private final Runnable allPlayersCompleted;
    private final List<SbModule> queue = new ArrayList<>();
    private final Object2LongMap<UUID> lastEdited = new Object2LongOpenHashMap<>();
    private final Set<UUID> edited = new HashSet<>();
    private final Set<UUID> completed = new HashSet<>();
    @Setter
    private boolean buildingPhase = false;
    private SbModule currentModule = null;
    @Setter
    private PlayerTeam team = null;
    private int successiveCompletion = 0;
    private int round = 0;

    public SbManager(Map<UUID, SbIsland> islands, List<SbModule> modules, MiniGameHandle gameHandle, ServerLevel world,
                     Random random, Runnable allPlayersCompleted) {
        this.islands = islands;
        this.modules = Collections.unmodifiableList(modules);
        this.gameHandle = gameHandle;
        this.logger = gameHandle.getLogger();
        this.world = world;
        this.random = random;
        this.allPlayersCompleted = allPlayersCompleted;
    }

    public void eachIsland(BiConsumer<SbIsland, ServerPlayer> action) {
        PlayerList playerManager = gameHandle.getServer().getPlayerList();

        islands.forEach((uuid, island) -> {
            ServerPlayer player = playerManager.getPlayer(uuid);

            if (player != null) {
                action.accept(island, player);
            }
        });
    }

    public boolean canModify(ServerPlayer player) {
        return buildingPhase && !completed.contains(player.getUUID());
    }

    public void clearIslands() {
        for (var island : activeIslands()) {
            island.getValue().clearBuildingArea(world);
        }
    }

    public boolean isWithinBuildingArea(ServerPlayer player, BlockPos pos) {
        SbIsland island = islands.get(player.getUUID());

        return island != null && island.isWithinBuildingArea(pos);
    }

    public synchronized void setModule(SbModule module) {
        CustomScoreboardManager scoreboardManager = gameHandle.getScoreboardManager();
        PlayerList playerManager = gameHandle.getServer().getPlayerList();

        for (var entry : activeIslands()) {
            ServerPlayer player = playerManager.getPlayer(entry.getKey());

            if (player == null) continue;

            SbIsland island = entry.getValue();

            if (!island.supports(module)) {
                logger.error("Module {} is incompatible with island {}", module, island);
                continue;
            }

            island.clearBuildingArea(world);
            island.placeModulePreview(module, world, team, scoreboardManager);
        }

        currentModule = module;
    }

    @NotNull
    public SbModule nextModule() {
        if (queue.isEmpty()) {
            queue.addAll(modules);
            Collections.shuffle(queue, random);
        }

        if (queue.isEmpty()) {
            throw new IllegalStateException("There are no modules defined");
        }

        return queue.removeFirst();
    }

    public Optional<ServerPlayer> getWorstPlayer() {
        var evaluation = evaluate();
        var minScore = evaluation.values().stream().mapToInt(Integer::intValue).min().orElse(0);

        return evaluation.entrySet().stream()
                // find players with minScore
                .filter(entry -> entry.getValue() == minScore)
                // sort by last edited
                .sorted(Comparator.<Map.Entry<ServerPlayer, Integer>>comparingLong(entry ->
                        lastEdited.getOrDefault(entry.getKey().getUUID(), Long.MAX_VALUE)).reversed())
                // map to actual player
                .map(Map.Entry::getKey)
                .findFirst();
    }

    private Set<Map.Entry<UUID, SbIsland>> activeIslands() {
        Participants participants = gameHandle.getParticipants();

        return islands.entrySet().stream()
                .filter(entry -> participants.isParticipating(entry.getKey()))
                .collect(Collectors.toSet());
    }

    private Map<ServerPlayer, Integer> evaluate() {
        if (currentModule == null) {
            return Map.of();
        }

        PlayerList playerManager = gameHandle.getServer().getPlayerList();
        Map<ServerPlayer, Integer> scores = new HashMap<>();

        for (var entry : activeIslands()) {
            ServerPlayer player = playerManager.getPlayer(entry.getKey());

            if (player == null) continue;

            SbIsland island = entry.getValue();
            int score = island.evaluate(world, currentModule);

            scores.put(player, score);
        }

        return scores;
    }

    public List<? extends Entity> getPreviewEntities() {
        var it = activeIslands().iterator();

        if (!it.hasNext()) {
            return List.of();
        }

        return it.next().getValue().getPreviewEntities(world);
    }

    public void onEdit(ServerPlayer player) {
        if (currentModule == null || completed.contains(player.getUUID())) return;

        lastEdited.put(player.getUUID(), System.currentTimeMillis());

        edited.add(player.getUUID());
    }

    public void tick() {
        checkPlayerPositions();
        processEdits();
    }

    private void processEdits() {
        if (edited.isEmpty()) return;

        Participants participants = gameHandle.getParticipants();
        PlayerList playerManager = gameHandle.getServer().getPlayerList();

        for (UUID uuid : edited) {
            if (!participants.isParticipating(uuid)) continue;

            ServerPlayer player = playerManager.getPlayer(uuid);

            if (player == null) continue;

            onEdited(player);
        }

        edited.clear();
    }

    private void checkPlayerPositions() {
        PlayerList playerManager = gameHandle.getServer().getPlayerList();

        for (var entry : activeIslands()) {
            ServerPlayer player = playerManager.getPlayer(entry.getKey());

            if (player == null || !player.getAbilities().mayfly) continue;

            SbIsland island = entry.getValue();

            if (island.getMovementBounds().contains(player.position())) continue;

            island.teleport(player);
        }
    }

    private void onEdited(ServerPlayer player) {
        // check if the player's used all the materials
        Inventory inventory = player.getInventory();

        for (int i = 0, size = inventory.getContainerSize(); i < size; i++) {
            ItemStack stack = inventory.getItem(i);

            if (stack.isEmpty() || stack.is(Items.WATER_BUCKET) || stack.is(Items.LAVA_BUCKET)) continue;

            return;
        }

        logger.info("Player {} has no items left", player.getScoreboardName());

        // the player used all the materials, check if the building is complete
        SbIsland island = islands.get(player.getUUID());

        if (island == null || !island.isCompleted(world, currentModule)) return;

        if (!completed.add(player.getUUID())) return;

        logger.info("Player {} has completed the building", player.getScoreboardName());

        ServerPlayerAccess.playSoundToPlayer(player, SoundEvents.PLAYER_LEVELUP, SoundSource.PLAYERS, 0.75f, 1.1f);

        var msg = gameHandle.getTranslations().translateText(player, "game.ap2.speed_builders.completed")
                .formatted(ChatFormatting.GREEN);

        player.sendSystemMessage(msg);

        checkOverallCompletion();
    }

    private void checkOverallCompletion() {
        if (completed.size() < gameHandle.getParticipants().count()) return;

        logger.info("All players completed their buildings");
        allPlayersCompleted.run();
    }

    public Optional<SbIsland> getIsland(ServerPlayer player) {
        return Optional.ofNullable(islands.get(player.getUUID()));
    }

    public boolean allIslandsComplete() {
        PlayerList playerManager = gameHandle.getServer().getPlayerList();

        return islands.entrySet().stream().allMatch(entry -> {
            UUID uuid = entry.getKey();
            ServerPlayer player = playerManager.getPlayer(uuid);

            if (player == null) {
                // do not count offline players
                return true;
            }

            SbIsland island = entry.getValue();

            return island.isCompleted(world, currentModule);
        });
    }

    public void incrementSuccessiveCompletion() {
        successiveCompletion++;
    }

    public void resetSuccessiveCompletion() {
        successiveCompletion = 0;
    }

    public void reset() {
        this.completed.clear();
        this.lastEdited.clear();
        this.edited.clear();
    }

    public int getBuildingDurationTicks() {
        if (currentModule == null) {
            return BASE_BUILD_DURATION_SECONDS;
        }

        int complexity = currentModule.getComplexity();
        int bonusTime = (int) Math.floor((Math.max(0, complexity - 64) * 0.4));
        int reduction = Math.max(0, successiveCompletion * SUCCESSIVE_COMPLETION_REDUCTION_SECONDS);

        return Math.max(MIN_BUILD_DURATION_SECONDS, BASE_BUILD_DURATION_SECONDS + bonusTime - reduction);
    }

    public void incrementRound() {
        round++;
    }

    public int getRoundsCompleted(ServerPlayer player, boolean winner) {
        if (completed.contains(player.getUUID()) || winner) {
            return round + 1;
        }

        return round;
    }
}

File: src/minigames/speed_builders/src/main/kotlin/work/lclpnet/ap2/game/speed_builders/util/SbManager.kt
package work.lclpnet.ap2.game.speed_builders.util

import it.unimi.dsi.fastutil.objects.Object2LongOpenHashMap
import net.minecraft.ChatFormatting
import net.minecraft.core.BlockPos
import net.minecraft.server.level.ServerLevel
import net.minecraft.server.level.ServerPlayer
import net.minecraft.sounds.SoundEvents
import net.minecraft.sounds.SoundSource
import net.minecraft.world.entity.Entity
import net.minecraft.world.entity.player.Inventory
import net.minecraft.world.item.Items
import net.minecraft.world.scores.PlayerTeam
import work.lclpnet.ap2.ext.mc.isOf
import work.lclpnet.ap2.game.MiniGameHandle
import work.lclpnet.ap2.game.speed_builders.data.SbIsland
import work.lclpnet.ap2.game.speed_builders.data.SbModule
import work.lclpnet.kibu.access.entity.ServerPlayerAccess
import java.util.*
import kotlin.math.floor
import kotlin.time.Duration
import kotlin.time.Duration.Companion.seconds
import kotlin.time.times

private val BASE_BUILD_DURATION = 25.seconds
private val MIN_BUILD_DURATION = 5.seconds
private val SUCCESSIVE_COMPLETION_REDUCTION = 7.seconds

class SbManager(
    private val islands: Map<UUID, SbIsland>,
    modules: List<SbModule>,
    private val gameHandle: MiniGameHandle,
    private val world: ServerLevel,
    private val random: Random,
    private val fastMode: Boolean,
    private val allPlayersCompleted: Runnable,
    private val lastPlayerRemaining: Runnable
) {
    private val modules: List<SbModule> = Collections.unmodifiableList(modules)
    private val logger = gameHandle.logger
    private val queue = mutableListOf<SbModule>()
    private val lastEdited = Object2LongOpenHashMap<UUID>()
    private val edited = mutableSetOf<UUID>()
    private val completed = mutableSetOf<UUID>()
    var buildingPhase: Boolean = false
    private var currentModule: SbModule? = null
    var team: PlayerTeam? = null
    private var successiveCompletion = 0
    private var round = 0
    private var roundResolved = false

    fun eachIsland(action: (SbIsland, ServerPlayer) -> Unit) {
        val playerManager = gameHandle.server.playerList
        islands.forEach { (uuid, island) ->
            val player = playerManager.getPlayer(uuid) ?: return@forEach
            action(island, player)
        }
    }

    fun canModify(player: ServerPlayer): Boolean = buildingPhase && !completed.contains(player.uuid)

    fun clearIslands() {
        for ((_, island) in activeIslands()) {
            island.clearBuildingArea(world)
        }
    }

    fun isWithinBuildingArea(player: ServerPlayer, pos: BlockPos): Boolean =
        islands[player.uuid]?.isWithinBuildingArea(pos) ?: false

    @Synchronized
    fun setModule(module: SbModule) {
        val scoreboardManager = gameHandle.scoreboardManager
        val playerManager = gameHandle.server.playerList

        for ((uuid, island) in activeIslands()) {
            playerManager.getPlayer(uuid) ?: continue

            if (!island.supports(module)) {
                logger.error("Module {} is incompatible with island {}", module, island)
                continue
            }

            island.clearBuildingArea(world)
            island.placeModulePreview(module, world, team!!, scoreboardManager)
        }

        currentModule = module
    }

    fun nextModule(): SbModule {
        if (queue.isEmpty()) {
            queue.addAll(modules)
            queue.shuffle(random)
        }

        if (queue.isEmpty()) throw IllegalStateException("There are no modules defined")

        return queue.removeFirst()
    }

    fun getWorstPlayer(): ServerPlayer? {
        val evaluation = evaluate()
        if (evaluation.isEmpty()) return null

        val minScore = evaluation.values.minOrNull() ?: return null

        return evaluation.entries
            .filter { (_, score) -> score == minScore }
            .maxByOrNull { (player, _) -> lastEdited.getOrDefault(player.uuid, Long.MAX_VALUE) }
            ?.key
    }

    private fun activeIslands(): Set<Map.Entry<UUID, SbIsland>> {
        val participants = gameHandle.participants
        return islands.entries.filter { (uuid, _) -> participants.isParticipating(uuid) }.toSet()
    }

    private fun evaluate(): Map<ServerPlayer, Int> {
        val module = currentModule ?: return emptyMap()
        val playerManager = gameHandle.server.playerList

        return activeIslands().mapNotNull { (uuid, island) ->
            val player = playerManager.getPlayer(uuid) ?: return@mapNotNull null
            player to island.evaluate(world, module)
        }.toMap()
    }

    fun getPreviewEntities(): List<Entity> =
        activeIslands().firstOrNull()?.value?.getPreviewEntities(world) ?: emptyList()

    fun onEdit(player: ServerPlayer) {
        if (currentModule == null || completed.contains(player.uuid)) return
        lastEdited.put(player.uuid, System.currentTimeMillis())
        edited.add(player.uuid)
    }

    fun tick() {
        checkPlayerPositions()
        processEdits()
    }

    private fun processEdits() {
        if (edited.isEmpty()) return

        val participants = gameHandle.participants
        val playerManager = gameHandle.server.playerList

        for (uuid in edited) {
            if (!participants.isParticipating(uuid)) continue
            val player = playerManager.getPlayer(uuid) ?: continue
            onEdited(player)
        }

        edited.clear()
    }

    private fun checkPlayerPositions() {
        val playerManager = gameHandle.server.playerList

        for ((uuid, island) in activeIslands()) {
            val player = playerManager.getPlayer(uuid) ?: continue
            if (!player.abilities.mayfly) continue
            if (island.movementBounds.contains(player.position())) continue
            island.teleport(player)
        }
    }

    private fun onEdited(player: ServerPlayer) {
        val inventory: Inventory = player.inventory

        for (i in 0 until inventory.containerSize) {
            val stack = inventory.getItem(i)
            if (stack.isEmpty || stack.isOf(Items.WATER_BUCKET) || stack.isOf(Items.LAVA_BUCKET)) continue
            return
        }

        logger.debug("Player {} has no items left", player.scoreboardName)

        val island = islands[player.uuid] ?: return
        if (!island.isCompleted(world, currentModule!!)) return
        if (!completed.add(player.uuid)) return

        logger.debug("Player {} has completed the building", player.scoreboardName)

        ServerPlayerAccess.playSoundToPlayer(player, SoundEvents.PLAYER_LEVELUP, SoundSource.PLAYERS, 0.75f, 1.1f)

        val msg = gameHandle.translations.translateText(player, "completed")
            .formatted(ChatFormatting.GREEN)
        player.sendSystemMessage(msg)

        checkOverallCompletion()
    }

    private fun checkOverallCompletion() {
        if (roundResolved) return

        val participantCount = gameHandle.participants.count()

        if (completed.size >= participantCount) {
            logger.debug("All players completed their buildings")
            roundResolved = true
            allPlayersCompleted.run()
            return
        }

        if (fastMode && completed.size >= participantCount - 1) {
            logger.debug("Only one builder remaining, eliminating instantly (fast mode)")
            roundResolved = true
            lastPlayerRemaining.run()
        }
    }

    fun getIsland(player: ServerPlayer): SbIsland? = islands[player.uuid]

    fun allIslandsComplete(): Boolean {
        val playerManager = gameHandle.server.playerList
        return islands.all { (uuid, island) ->
            playerManager.getPlayer(uuid) ?: return@all true

            island.isCompleted(world, currentModule!!)
        }
    }

    fun incrementSuccessiveCompletion() { successiveCompletion++ }
    fun resetSuccessiveCompletion() { successiveCompletion = 0 }

    fun reset() {
        completed.clear()
        lastEdited.clear()
        edited.clear()
        roundResolved = false
    }

    fun getBuildingDuration(): Duration {
        val module = currentModule ?: return BASE_BUILD_DURATION
        val complexity = module.getComplexity()
        val bonusTime = floor((maxOf(0, complexity - 64) * 0.4)).seconds
        val reduction = maxOf(0.seconds, successiveCompletion * SUCCESSIVE_COMPLETION_REDUCTION)
        return maxOf(MIN_BUILD_DURATION, BASE_BUILD_DURATION + bonusTime - reduction)
    }

    fun incrementRound() { round++ }

    fun getRoundsCompleted(player: ServerPlayer, winner: Boolean): Int =
        if (completed.contains(player.uuid) || winner) round + 1 else round
}
